File: inference.py
import os

os.environ["KERAS_BACKEND"] = "tensorflow"

import keras
import json
import numpy as np
import random
from tabulate import tabulate
import tensorflow as tf

print(f"TF Version: {tf.__version__}")
print(f"Keras Version: {keras.__version__}")
print("Physical Devices:", tf.config.list_physical_devices())
print("Using oneDNN:", tf.sysconfig.get_build_info().get('is_cuda_build'))

# 1. SETUP & LOAD
MAX_LEN = 40  # Must match the value used in training!

# Load the trained model
print("Loading model...")
model = keras.models.load_model("best_poke_model.keras") #, safe_mode=False)

@tf.function(reduce_retracing=True)
def model_step(enc_tensor, dec_input):
    return model([enc_tensor, dec_input], training=False)


# Load the vocabulary
with open("vocab.json", "r") as f:
    vocab = json.load(f)

# Invert vocab (ID -> Char) for decoding
inv_vocab = {v: k for k, v in vocab.items()}

# Special Token IDs
START_TOKEN = vocab["["]
STOP_TOKEN = vocab["]"]
PAD_TOKEN = vocab["[PAD]"]


# Optimized Beam Search with Batching
def decode_sequence_beam_batched(input_text, task_token, beam_width=3):
    full_text = task_token + input_text.lower()
    enc_ids = [vocab.get(c, 0) for c in full_text]
    enc_ids = enc_ids[:MAX_LEN]
    enc_ids += [PAD_TOKEN] * (MAX_LEN - len(enc_ids))

    # Shape: (beam_width, MAX_LEN) -> e.g. (3, 40)
    enc_tensor = tf.convert_to_tensor([enc_ids] * beam_width, dtype=tf.int32)

    # 2. INITIALIZE BEAMS
    # Beams are now kept as parallel lists/arrays
    # If we start all scores at 0, we get 3 identical beams.
    scores = tf.constant([0.0] + [-1e9] * (beam_width - 1))

    # Sequences: (beam_width, 1) -> [[START], [START], [START]]
    sequences = tf.constant([[START_TOKEN]] * beam_width, dtype=tf.int32)

    # Track which beams have finished
    finished_seqs = []
    finished_scores = []

    # 3. AUTOREGRESSIVE LOOP
    for i in range(MAX_LEN - 1):
        # A. PREPARE DECODER INPUT
        # Pad current sequences to (beam_width, MAX_LEN-1)
        # We need manual padding here because TF tensor shapes are strict
        curr_len = sequences.shape[1]
        pad_size = (MAX_LEN - 1) - curr_len

        if pad_size > 0:
            paddings = tf.constant([[0, 0], [0, pad_size]])  # (batch, time)
            dec_input = tf.pad(sequences, paddings, constant_values=PAD_TOKEN)
        else:
            dec_input = sequences

        # B. BATCHED INFERENCE (One call for all beams!)
        # Output Shape: (beam_width, MAX_LEN-1, vocab_size)
        preds = model_step(enc_tensor, dec_input)

        # Get logits for the last token only: (beam_width, vocab_size)
        next_token_logits = preds[:, curr_len - 1, :]

        # Convert to Log Probs: (beam_width, vocab_size)
        log_probs = tf.nn.log_softmax(next_token_logits)

        # C. EXPAND BEAMS
        # Add current beam scores to the new log probs
        # score[b] + log_prob[b, v]
        # Shape: (beam_width, vocab_size)
        candidate_scores = tf.expand_dims(scores, axis=1) + log_probs

        # Flatten to find top K across ALL beams * ALL vocab
        # Shape: (beam_width * vocab_size,)
        flat_scores = tf.reshape(candidate_scores, [-1])

        # Top K scores and indices
        top_k_scores, top_k_indices = tf.math.top_k(flat_scores, k=beam_width)

        # D. RECONSTRUCT BEAMS
        vocab_size = log_probs.shape[-1]
        beam_indices = top_k_indices // vocab_size
        token_indices = top_k_indices % vocab_size

        next_sequences = []
        next_scores = []

        # Update sequences
        for k in range(beam_width):
            beam_idx = beam_indices[k]
            token = token_indices[k]
            score = top_k_scores[k]

            # If token is STOP, move to finished list
            if token == STOP_TOKEN:
                finished_seqs.append(sequences[beam_idx])
                finished_scores.append(score)
                dummy_seq = tf.concat([sequences[beam_idx], [PAD_TOKEN]], axis=0)
                next_sequences.append(dummy_seq)
                next_scores.append(-1e9) # Kill this beam
            else:
                new_seq = tf.concat([sequences[beam_idx], [token]], axis=0)
                next_sequences.append(new_seq)
                next_scores.append(score)
        # Stack back into tensors for next loop
        sequences = tf.stack(next_sequences)  # (beam_width, seq_len + 1)
        scores = tf.stack(next_scores)

        # Early Exit: If all active beams are terrible (very low score), stop
        if tf.reduce_max(scores) < -1e8:
            print("EARLY BREAK!", i)
            break

    # 4. SELECT BEST
    # If we have finished sequences, pick best.
    # If loop finished without STOP token, pick best active beam.
    if len(finished_scores) > 0:
        # Apply penalty: score / (len^alpha)
        # Using a simple length normalization for starters:
        # TODO: alpha as a parameter?
        normalized_scores = [
            s / (len(seq)**0.7) for s, seq in zip(finished_scores, finished_seqs)
        ]
        best_idx = np.argmax(normalized_scores)
        best_seq = finished_seqs[best_idx].numpy()
    else:
        best_idx = np.argmax(scores)
        best_seq = sequences[best_idx].numpy()

    # Decode
    decoded_chars = []
    for idx in best_seq:
        if idx == START_TOKEN:
            continue
        if idx == STOP_TOKEN:
            break
        decoded_chars.append(inv_vocab.get(idx, ""))

    return "".join(decoded_chars)


# Updated wrapper
def generate_ipa(word):
    return decode_sequence_beam_batched(word, "<", beam_width=5)

def generate_word(ipa):
    return decode_sequence_beam_batched(ipa, ">", beam_width=5)


def load_random_pokemon(file_path, n=10):
    with open(file_path, "r", encoding="utf-8") as f:
        # Filter out empty lines
        lines = [line.strip() for line in f if line.strip()]

    # Sample first, then parse (more efficient)
    samples = random.sample(lines, min(n, len(lines)))

    pairs = []
    for line in samples:
        parts = line.split("\t")
        if len(parts) < 2:
            continue  # Skip bad lines

        word, ipa = parts[0], parts[1]

        # --- THE FIX ---
        # Remove slashes '/' and strip whitespace
        ipa = ipa.replace("/", "").strip()

        pairs.append((word, ipa))

    return pairs


samples = load_random_pokemon("pokemon.tsv", n=10)

rows = []

for word, true_ipa in samples:
    try:
        print("Processing", word, "...")
        model_ipa = generate_ipa(word)
        model_word = generate_word(true_ipa)
        round_robin = generate_word(model_ipa)

        rows.append([word, f"{true_ipa}", f"{model_ipa}", model_word, round_robin])
    except Exception as e:
        rows.append([word, f"Error: {e}", "", "", ""])

# Print table nicely
headers = ["POKEMON", "TRUE IPA", "WORD→IPA", "IPA→WORD", "ROUND-TRIP"]
print(tabulate(rows, headers=headers, tablefmt="fancy_grid"))

import gc
gc.collect()
tf.keras.backend.clear_session()

File: inference.js
import * as tf from 'https://esm.run/@tensorflow/tfjs';
import {SequenceScorer} from './shallow_fusion.js';

// TODO: Include this in the model context?
const MAX_LEN = 40;

export async function decodeBeamBatched(
  model,
  vocab,
  invVocab,
  word,
  taskToken,
  beam_width = 3,
  temperature = 1.0,
  alpha = 0.6,
  logitsProcessors = [], // Array of { processor: LogitsProcessor, weight: number }
  sequenceScorers = []   // Array of { scorer: SequenceScorer, weight: number }
) {
  const fusionContext = { originalInput: word, taskToken, vocab, invVocab };

  // PHASE 1: Generate Candidates (Forward Pass)
  const candidates = tf.tidy(() => {
    let cleanInput = word;
    if (taskToken === "<") {
      cleanInput = word.toLowerCase();
    }
    const fullText = taskToken + cleanInput;
    const START_TOKEN = vocab["["];
    const STOP_TOKEN = vocab["]"];
    const PAD_TOKEN = vocab["[PAD]"];

    const getAlphaScore = (logProb, length) => {
      const lp = Math.pow((5 + length) / 6, alpha);
      return logProb / lp;
    };

    let encIds = fullText
      .split("")
      .map((c) => vocab[c] || 0)
      .slice(0, MAX_LEN);
    while (encIds.length < MAX_LEN) encIds.push(PAD_TOKEN);

    const encTensor = tf.tensor2d(
      Array(beam_width).fill(encIds),
      [beam_width, MAX_LEN],
      "float32",
    );
    let scores = tf.tensor1d([0.0, ...Array(beam_width - 1).fill(-1e9)]);
    let sequences = tf.fill([beam_width, 1], START_TOKEN, "int32");

    let finishedSeqs = [];
    let finishedScores = [];

    // Tracking arrays for LogitsProcessors
    let activeProcessorTracking = Array.from({ length: beam_width }, () => ({}));
    let finishedProcessorTracking = [];

    for (let i = 0; i < MAX_LEN - 1; i++) {
      const currLen = sequences.shape[1];
      const padSize = MAX_LEN - 1 - currLen;
      let decInput = sequences.cast("float32");
      if (padSize > 0)
        decInput = decInput.pad(
          [
            [0, 0],
            [0, padSize],
          ],
          PAD_TOKEN,
        );

      const inputs = { enc_in: encTensor, dec_in: decInput };
      let preds = model.execute(inputs);
      if (Array.isArray(preds)) preds = preds[0];

      const nextTokenLogits = preds
        .gather([currLen - 1], 1)
        .reshape([beam_width, -1]);

      // Use let so we can modify it
      let logProbs = tf.log(nextTokenLogits.add(1e-9));

      // >>> 1. APPLY TEMPERATURE SCALING >>>
      if (temperature !== 1.0) {
        const scaledLogProbs = tf.tidy(() => {
          const scaled = logProbs.div(temperature);
          return tf.logSoftmax(scaled);
        });

        logProbs.dispose(); // Prevent memory leak
        logProbs = scaledLogProbs;
      }

      // >>> 2. SHALLOW FUSION: LOGITS PROCESSING >>>
      const stepDeltas = {};

      for (const { processor, weight } of logitsProcessors) {
        if (weight === 0) continue;

        const deltaInfo = tf.tidy(() => {
          const vocabSize = logProbs.shape[1];
          const deltaTensor = processor.process(sequences.arraySync(), vocabSize, fusionContext);

          if (deltaTensor) {
            const weightedDelta = deltaTensor.mul(weight);
            return {
              tensor: weightedDelta,
              jsArr: weightedDelta.arraySync() // Extract values for tracking
            };
          }
          return null;
        });

        if (deltaInfo) {
          stepDeltas[processor.name] = deltaInfo.jsArr;

          const newLogProbs = logProbs.add(deltaInfo.tensor);
          logProbs.dispose();
          logProbs = newLogProbs;
          deltaInfo.tensor.dispose();
        }
      }

      const candidateScores = scores.expandDims(1).add(logProbs);
      const flatScores = candidateScores.reshape([-1]);
      const { values: topKScores, indices: topKIndices } = tf.topk(
        flatScores,
        beam_width,
      );

      const vocabSize = logProbs.shape[1];
      const beamIndices = topKIndices
        .div(tf.scalar(vocabSize, "int32"))
        .cast("int32");
      const tokenIndices = topKIndices
        .mod(tf.scalar(vocabSize, "int32"))
        .cast("int32");

      const bIdxArr = beamIndices.arraySync();
      const tIdxArr = tokenIndices.arraySync();
      const sArr = topKScores.arraySync();
      const seqsArr = sequences.arraySync();

      const nextSeqs = [];
      const nextScoresArr = [];
      const nextProcessorTracking = [];

      for (let k = 0; k < beam_width; k++) {
        const bIdx = bIdxArr[k];
        const token = tIdxArr[k];
        const score = sArr[k];

        // Track processor influence for this specific branch
        const newTrack = { ...activeProcessorTracking[bIdx] };
        for (const pName in stepDeltas) {
          const appliedScore = stepDeltas[pName][bIdx][token];
          newTrack[pName] = (newTrack[pName] || 0) + appliedScore;
        }

        if (token === STOP_TOKEN) {
          finishedSeqs.push(seqsArr[bIdx]);
          finishedScores.push(score);
          finishedProcessorTracking.push(newTrack);

          nextSeqs.push([...seqsArr[bIdx], PAD_TOKEN]);
          nextScoresArr.push(-1e9);
          nextProcessorTracking.push({});
        } else {
          nextSeqs.push([...seqsArr[bIdx], token]);
          nextScoresArr.push(score);
          nextProcessorTracking.push(newTrack);
        }
      }

      sequences = tf.tensor2d(
        nextSeqs,
        [beam_width, nextSeqs[0].length],
        "int32",
      );
      scores = tf.tensor1d(nextScoresArr);
      activeProcessorTracking = nextProcessorTracking; // Update state

      if (finishedSeqs.length >= beam_width) {
        const finishedWithAlpha = finishedScores.map((s, idx) => {
          const len = finishedSeqs[idx].length - 1;
          return getAlphaScore(s, len);
        });
        finishedWithAlpha.sort((a, b) => b - a);
        const worstWinningScore = finishedWithAlpha[beam_width - 1];
        const bestActiveLogProb = Math.max(...nextScoresArr);

        if (bestActiveLogProb > -1e8) {
          const maxPossibleLP = Math.pow((5 + MAX_LEN) / 6, alpha);
          if (bestActiveLogProb / maxPossibleLP < worstWinningScore) break;
        } else {
          break;
        }
      }
    }

    const finalActiveSeqs = sequences.arraySync();
    const finalActiveScores = scores.arraySync();
    let allCandidates = [];

    for (let i = 0; i < finishedSeqs.length; i++) {
      allCandidates.push({
        seq: finishedSeqs[i],
        rawScore: finishedScores[i],
        status: "DONE",
        processorTracking: finishedProcessorTracking[i]
      });
    }
    for (let i = 0; i < finalActiveSeqs.length; i++) {
      if (finalActiveScores[i] > -1e8) {
        allCandidates.push({
          seq: finalActiveSeqs[i],
          rawScore: finalActiveScores[i],
          status: "MAX_LEN",
          processorTracking: activeProcessorTracking[i]
        });
      }
    }

    return allCandidates
      .map((c) => {
        const cleanIds = c.seq.filter(
          (id) => id !== START_TOKEN && id !== STOP_TOKEN && id !== PAD_TOKEN,
        );
        const text = cleanIds.map((id) => invVocab[id]).join("");
        const alphaScore = getAlphaScore(c.rawScore, cleanIds.length);

        return {
          Text: text,
          "Alpha Score": alphaScore,
          "Raw LogProb": c.rawScore,
          Status: c.status,
          processorTracking: c.processorTracking // Pass tracking data through
        };
      })
      .sort((a, b) => b["Alpha Score"] - a["Alpha Score"]);
  }); // End tf.tidy

  // --- PHASE 2 & 3: Sequence Scorers & Reranking ---

  // 1. Initialize Fused Score with the Forward Alpha Score
  candidates.forEach(c => {
    c["Fused Score"] = parseFloat(c["Alpha Score"]);
  });

  // 2. Apply each Sequence Scorer dynamically
  for (const { scorer, weight } of sequenceScorers) {
    if (weight === 0) continue;

    const finishScores = await scorer.score(candidates, fusionContext);

    if (finishScores) {
      for (let i = 0; i < candidates.length; i++) {
        const scoreDelta = weight * finishScores[i];
        candidates[i]["Fused Score"] += scoreDelta;
        candidates[i][`ScorerRaw_${scorer.name}`] = finishScores[i].toFixed(4);
      }
    }
  }

  // 3. Re-sort candidates based on the new Fused Score (Descending)
  candidates.sort((a, b) => b["Fused Score"] - a["Fused Score"]);

  // 4. Build and log the detailed reranking table dynamically
  const tableData = candidates.map((row, idx) => {
    const tableRow = {
      "Rank": idx + 1,
      "Text": row.Text,
      "Fused Score": row["Fused Score"].toFixed(4),
      "Fwd Alpha": row["Alpha Score"].toFixed(4),
      "Status": row.Status,
    };

    // Dynamically append columns for Logits Processors
    if (row.processorTracking) {
      for (const [pName, accumulatedScore] of Object.entries(row.processorTracking)) {
        const config = logitsProcessors.find(p => p.processor.name === pName);
        const w = config ? config.weight : "?";
        tableRow[`LP: ${pName} (w=${w})`] = accumulatedScore.toFixed(4);
      }
    }

    // Dynamically append columns for Sequence Scorers
    for (const { scorer, weight } of sequenceScorers) {
      if (weight !== 0 && row[`ScorerRaw_${scorer.name}`] !== undefined) {
        tableRow[`SS: ${scorer.name} (w=${weight})`] = row[`ScorerRaw_${scorer.name}`];
      }
    }

    return tableRow;
  });

  const activeStr = [
    ...logitsProcessors.filter(p => p.weight > 0).map(p => `LP_${p.processor.name}=${p.weight}`),
    ...sequenceScorers.filter(s => s.weight > 0).map(s => `SS_${s.scorer.name}=${s.weight}`)
  ].join(" | ");

  console.log(
    `%cBeam Search (${taskToken}) | ${activeStr || "Vanilla Keras"}`,
    "font-weight: bold; color: #4CAF50; font-size: 12px;",
  );
  console.table(tableData);

  // We now return the top-ranked item after fusion.
  return candidates.length > 0 ? candidates[0].Text : "";
}

export async function computeCyclicLossBatch(
  model,
  vocab,
  invVocab,
  generatedCandidates,
  originalInput,
  backwardTaskToken,
) {
    // TODO: Can we use the length of the originalInput here rather than MAX_LEN?
  return tf.tidy(() => {
    const BATCH_SIZE = generatedCandidates.length;
    const PAD_TOKEN = vocab["[PAD]"];

    // --- 1. Prepare Encoder Inputs ---
    const encIdsBatch = generatedCandidates.map((txt) => {
      let cleanTxt = txt;
      if (backwardTaskToken === "<") {
        cleanTxt = txt.toLowerCase();
      }
      const encText = backwardTaskToken + cleanTxt;
      let ids = encText
        .split("")
        .map((c) => vocab[c] || 0)
        .slice(0, MAX_LEN);
      while (ids.length < MAX_LEN) ids.push(PAD_TOKEN);
      return ids;
    });
    const encTensor = tf.tensor2d(
      encIdsBatch,
      [BATCH_SIZE, MAX_LEN],
      "float32",
    );

    // --- 2. Prepare Decoder Inputs ---
    let targetTxt = originalInput;
    if (backwardTaskToken === ">") {
      targetTxt = originalInput.toLowerCase();
    }

    const fullTargetText = "[" + targetTxt + "]";
    let allTargetIds = fullTargetText.split("").map((c) => vocab[c] || 0);

    let decInputIds = allTargetIds.slice(0, MAX_LEN - 1);
    while (decInputIds.length < MAX_LEN - 1) decInputIds.push(PAD_TOKEN);

    const decInputSingle = tf.tensor2d(
      [decInputIds],
      [1, MAX_LEN - 1],
      "float32",
    );
    const decInputBatch = decInputSingle.tile([BATCH_SIZE, 1]);

    // --- 3. Run Model ---
    let preds = model.execute({ enc_in: encTensor, dec_in: decInputBatch });
    if (Array.isArray(preds)) preds = preds[0];

    // --- 4. Compute Loss ---
    const results = [];
    const limit = Math.min(allTargetIds.length - 1, MAX_LEN - 1);

    for (let b = 0; b < BATCH_SIZE; b++) {
      // Get prediction matrix for this batch item: [SEQ_LEN, VOCAB_SIZE]
      const batchPreds = preds.gather([b], 0).squeeze([0]);

      let totalLoss = 0;
      const charDetails = [];

      for (let i = 0; i < limit; i++) {
        const nextCharId = allTargetIds[i + 1];

        const charProb = batchPreds
          .slice([i, nextCharId], [1, 1])
          .dataSync()[0];

        const safeProb = Math.max(charProb, 1e-9);
        const stepLoss = -Math.log(safeProb);

        totalLoss += stepLoss;

        charDetails.push({
          char: invVocab[nextCharId],
          prob: (charProb * 100).toFixed(2) + "%",
          loss: stepLoss.toFixed(4),
        });
      }
      results.push({
        avgLoss: (totalLoss / limit).toFixed(4),
        details: charDetails,
      });
    }
    return results;
  });
}

export async function computeCyclicLoss(
  model,
  vocab,
  invVocab,
  generatedIpa,
  originalWord,
  backwardTaskToken,
) {
  const results = await computeCyclicLossBatch(
    model,
    vocab,
    invVocab,
    [generatedIpa],
    originalWord,
    backwardTaskToken,
  );
  return results[0];
}


// The Cyclic Loss model ONLY scores finished sequences.
// TODO: Link up properly.
export class CyclicSequenceScorer extends SequenceScorer {
  constructor(model, vocab, invVocab, computeCyclicLossBatchFn) {
    super("CyclicLoss");
    this.model = model;
    this.vocab = vocab;
    this.invVocab = invVocab;
    this.computeLoss = computeCyclicLossBatchFn;
  }

  async score(candidates, context) {
    const backwardTaskToken = context.taskToken === "<" ? ">" : "<";
    const textInputs = candidates.map(c => c.Text);

    const lossResults = await this.computeLoss(this.model, this.vocab, this.invVocab, textInputs, context.originalInput, backwardTaskToken);
    return lossResults.map(r => -parseFloat(r.avgLoss));
  }
}